fish-rust/src/builtins/argparse.rs

```diff
@@ -0,0 +1,1033 @@
+                }
+                woption_argument_t::required_argument
+            }
+            ArgCardinality::None => woption_argument_t::no_argument,
+        };
+
+        if !opt_spec.long_flag.is_empty() {
+            long_options.push(wopt(opt_spec.long_flag, arg_type, opt_spec.short_flag));
+        }
+    }
+}
+
+#[allow(unused_variables)]
+#[allow(unreachable_code)]
+fn validate_arg<'opts>(
+    parser: &mut parser_t,
+    opts_name: &wstr,
+    opt_spec: &mut OptionSpec<'opts>,
+    is_long_flag: bool,
+    woptarg: &'opts wstr,
+    streams: &mut io_streams_t,
+) -> Option<c_int> {
+    // Obviously if there is no arg validation command we assume the arg is okay.
+    if opt_spec.validation_command.is_empty() {
+        return STATUS_CMD_OK;
+    }
+
+    parser.get_var_stack().pin().push(true);
+
+    let env_mode = EnvMode::LOCAL | EnvMode::EXPORT;
+    parser.set_var(
+        L!("_argparse_cmd"),
+        array::from_ref(&opts_name).as_slice(),
+        env_mode,
+    );
+    let flag_name = WString::from(VAR_NAME_PREFIX) + "name";
+    if is_long_flag {
+        parser.set_var(
+            flag_name,
+            array::from_ref(&opt_spec.long_flag).as_slice(),
+            env_mode,
+        );
+    } else {
+        parser.set_var(
+            flag_name,
+            array::from_ref(&wstr::from_char_slice(&[opt_spec.short_flag])).as_slice(),
+            env_mode,
+        );
+    }
+    parser.set_var(
+        WString::from(VAR_NAME_PREFIX) + "value",
+        array::from_ref(&woptarg).as_slice(),
+        env_mode,
+    );
+
+    let mut cmd_output = Vec::new();
+
+    let retval = exec_subshell(opt_spec.validation_command, parser, &mut cmd_output, false);
+
+    for output in cmd_output {
+        streams.err.append(output);
+        streams.err.append1('\n');
+    }
+    parser.get_var_stack().pin().pop();
+    return retval;
+}
+
+/// \return whether the option 'opt' is an implicit integer option.
+fn is_implicit_int(opts: &ArgParseCmdOpts, val: &wstr) -> bool {
+    if opts.implicit_int_flag == '\0' {
+        // There is no implicit integer option.
+        return false;
+    }
+
+    // We succeed if this argument can be parsed as an integer.
+    fish_wcstol(val).is_ok()
+}
+
+// Store this value under the implicit int option.
+fn validate_and_store_implicit_int<'args>(
+    parser: &mut parser_t,
+    opts: &mut ArgParseCmdOpts<'args>,
+    val: &'args wstr,
+    w: &mut wgetopter_t,
+    is_long_flag: bool,
+    streams: &mut io_streams_t,
+) -> Option<c_int> {
+    let opt_spec = opts.options.get_mut(&opts.implicit_int_flag).unwrap();
+    let retval = validate_arg(parser, &opts.name, opt_spec, is_long_flag, val, streams);
+
+    if retval != STATUS_CMD_OK {
+        return retval;
+    }
+
+    // It's a valid integer so store it and return success.
+    opt_spec.vals.clear();
+    opt_spec.vals.push(val.into());
+    opt_spec.num_seen += 1;
+    w.nextchar = L!("");
```

Comment:
This used to set `nextchar` to `nullptr`, and `nextchar` was private before this change, is there a method on `wgetopter_t` that should be used instead?

Reply:
No what you did (making it public) is fine.